sdk/opa.go

```diff
@@ -298,6 +298,96 @@ func newDecisionResult() (*DecisionResult, error) {
 	return result, nil
 }
 
+// Partial returns a named decision. This function is threadsafe.
+func (opa *OPA) Partial(ctx context.Context, options PartialOptions) (*DecisionResult, error) {
+
+	if options.Resolver == nil {
+		return nil, fmt.Errorf("ResultResolver must be specified")
+	}
+
+	m := metrics.New()
+	m.Timer(metrics.SDKDecisionEval).Start()
+
+	result, err := newDecisionResult()
+	if err != nil {
+		return nil, err
+	}
+
+	opa.mtx.Lock()
+	s := *opa.state
+	opa.mtx.Unlock()
+
+	record := server.Info{
+		DecisionID: result.ID,
+		Timestamp:  options.Now,
+		Input:      &options.Input,
+		Query:      options.Query,
+		Path:       options.Path,
+		Metrics:    m,
+	}
+
+	if record.Timestamp.IsZero() {
+		record.Timestamp = time.Now().UTC()
+	}
+
+	if record.Path == "" {
+		record.Path = *s.manager.Config.DefaultDecision
+	}
+
+	record.Txn, record.Error = s.manager.Store.NewTransaction(ctx, storage.TransactionParams{})
+
+	if record.Error == nil {
+		defer s.manager.Store.Abort(ctx, record.Txn)
+		var pq *rego.PartialQueries
+		pq, record.InputAST, record.Bundles, record.Error = partial(ctx, partialEvalArgs{
+			runtime:   s.manager.Info,
+			printHook: s.manager.PrintHook(),
+			compiler:  s.manager.GetCompiler(),
+			store:     s.manager.Store,
+			txn:       record.Txn,
+			now:       record.Timestamp,
+			query:     record.Query,
+			unknowns:  options.Unknowns,
+			input:     *record.Input,
+			m:         record.Metrics,
+		})
+		if record.Error == nil {
+			var recordResults interface{}
+			result.Result, recordResults, err = options.Resolver.ProcessResults(pq)
+			if err != nil {
+				return nil, err
+			}
+			record.Results = &recordResults
+		}
+	}
+
+	m.Timer(metrics.SDKDecisionEval).Stop()
+
+	if logger := logs.Lookup(s.manager); logger != nil {
+		if err := logger.Log(ctx, &record); err != nil {
+			return result, fmt.Errorf("decision log: %w", err)
+		}
+	}
+
+	return result, record.Error
+}
+
+// The first interface being returned is the type that will be used for further processing
+// The second interface is for the decision logger, most of the time these should be the same
+type ResultResolver interface {
```

Comment:
nit: maybe instead of `ResultResolver` we call this something else like `PartialQueryMapper` which is a bit more accurate, i.e., the interface provides a mechanism for mapping partial evaluation results (which are exposed as rego.PartialQuery)... let's first decide on whether we need the interface in the first place.

Reply:
Yes, I wasn't happy about the name, PartialQueryMapper is a much better name.  I'll wait until we decide on the interface before I make the change.

Reply:
As I was in the code for the refactor, I did make this name change.  Can still pull the entire thing out based on the outcome of the other thread.